### ISSUE TITLE
Implement all Ops in PyTorch (help welcome!)

### ISSUE THREAD

**Issue by ricardoV94** [pymc-devs/pytensor#821]
### Description

If you want to help implementing some of these Ops just leave a comment below saying which ones you are interested in. We'll give you some time to work on it (and then put it back up to grabs).

See the documentation for How to implement PyTorch Ops and tests: https://pytensor.readthedocs.io/en/latest/extending/creating_a_numba_jax_op.html

Example PR: #836 

See https://github.com/pymc-devs/pytensor/issues/821#issuecomment-2202258929 for suggestions on equivalent torch functions

### Tensor creation Ops
- [x] Alloc and AllocEmpty #836 
- [x] Arange #836 
- [x] Eye #877 
- [ ] ScalarFromTensor
- [ ] TensorFromScalar
- [x] Repeat #890
- [x] Unique #890
- [x] Sort / Argsort #897
- [ ] Tri

### Shape Ops
- [x] Dimshuffle - Done in #764 
- [x] Reshape  #926
- [x] Shape, Shape_i #926
- [x] SpecifyShape #926
- [x] Unbroadcast #926
- [x] Join #869 
- [ ] Split
- [x] MakeVector #926

### Math Ops
- [x] Elemwise - Done in #764 (but not complete! Specific cases will require a custom ScalarOp dispatch)
- [x] CAReduce (Sum, All, Any...) - Assigned to @HarshvirSandhu 
- [x] CumOp - #837 
- [x] Softmax, LogSoftmax and Grads in #846 
- [x] Dot #878 
- [x] BatchedDot #878
- [ ] Argmax

### Indexing Ops
- [ ] Subtensor  #910
- [ ] Inc/SetSubtensor #910
- [ ] AdvancedSubtensor[1]  #910
- [ ] AdvancedIncSubtensor[1]  #910

### Branching Ops
- [x] CheckAndRaise - Done in #764 
- [ ] Ifelse #940 
- [ ] ScalarLoop
- [ ] Scan
- [ ] OpFromGraph
- [ ] Blockwise

### Linalg Ops
- [ ] SVD #920
- [ ] Det #920
- [ ] Eig #920
- [ ] Eigh #920
- [ ] MatrixInverse #920
- [ ] MatrixPinv #920
- [ ] QRFull #920
- [ ] SLogDet #920
- [ ] BlockDiagonal #922
- [ ] Cholesky  #922
- [ ] Solve  #922
- [ ] SolveTriangular  #922

### SparseOps
- [ ] ... (to be filled)

### RandomVariable Ops
 - [ ] ... Need to figure out API differences

If you need an Op that's not in this list, comment below and we'll add it!

**Comment by HAKSOAT** [comment]
Hi there, I'm attending the hackathon at Pydata London.

I'd like to work on Softmax.

**Comment by ricardoV94** [comment]
> Hi there @ricardoV94, I'm attending the hackathon at Pydata London.
> 
> I'd like to work on Softmax.

Sure!

**Comment by HAKSOAT** [comment]
Thanks @ricardoV94 

I have been doing some reading around the docs, I noticed that #764 has the initial setup. I believe that contains ground work for adding other Ops, so I'd be following the PR and the Softmax Ops will likely use some code from it.

Am I thinking about this the right way?

**Comment by ricardoV94** [comment]
@HAKSOAT yes, you should be able to start a branch already from #764. All the functionality is done, except some tests are failing because they depend non-optionally on GPU. We should make that optional and get it merged soon enough, but you need not wait :)

**Comment by ricardoV94** [comment]
#764 is merged

**Comment by aibit0111** [comment]
Hello @ricardoV94, I would like to work on Reshape

**Comment by ricardoV94** [comment]
Go ahead. We'll link and lock the Op when you open a PR

**Comment by HAKSOAT** [comment]
Hi @ricardoV94, I have opened a PR for the Softmax Ops. I see it has been grouped with LogSoftmax and Grads, so I can update the PR to include them.

**Comment by HangenYuu** [comment]
Hi @ricardoV94, I will work on the `Dot` `Op` now.

**Comment by ricardoV94** [comment]
If someone wants to look through the codebase and populate the list of Ops above that would also be very helpful :)

**Comment by twaclaw** [comment]
I could help with the remaining `Tensor creation ops` to begin with. Let me know.

**Comment by ricardoV94** [comment]
Thanks @twaclaw, feel free to open a PR

**Comment by twaclaw** [comment]
I will have a look at `Repeat`, `Unique`, etc. during the weekend.

@ricardoV94, regarding `Unique`:
* is it desired to emulate the behaviour of `return_index=True` or rather throw  `NotImplementedError`? (`torch.unique` doesn't feature that param)
* is the `axis` param required? Your current  `JAX` Op implementation doesn't support a non `None` `axis`

**Comment by ricardoV94** [comment]
@twaclaw in general we want to support exactly the same functionality from the original Op. When that is not possible or too complicated raising `NotImplementedError` is fine. 

Regarding JAX, we probably can compile (JIT) any function that has unique in it because JAX can't handle dynamic shapes. So it's a bit moot whether we say we support axis or not, although the NotImplementedError could be removed and we could just dispatch to `jax.numpy.unique` rather straightforward. Feel free to open a PR for that if you want.

**Comment by HarshvirSandhu** [comment]
I'll be working on the indexing Ops now

**Comment by twaclaw** [comment]
> @twaclaw in general we want to support exactly the same functionality from the original Op. When that is not possible or too complicated raising `NotImplementedError` is fine.
> 
> Regarding JAX, we probably cannot compile (JIT) any function that has unique in it because JAX can't handle dynamic shapes. So it's a bit moot whether we say we support axis or not, although the NotImplementedError could be removed and we could just dispatch to `jax.numpy.unique` rather straightforward. Feel free to open a PR for that if you want.

@ricardoV94, regarding the JAX implementation of `unique`, one possible option would be to make the param `size` in `jax.numpy.unique` static. Anyhow, I think the current implementation might be broken (i.e.,  `lax_numpy` is undefined).

**Comment by ricardoV94** [comment]
I can't imagine many cases where I would know the size of the unique elements but not what/where they were? If I knew what/where they are I would just select them instead of using unique.

**Comment by twaclaw** [comment]
You are right, `unique` is not compatible with JIT. 
You don't need to know the exact size but guess it (maybe overestimate it).  I was wondering whether it is possible to insert parameters (like a Constant `size`) into the Op Graph 🤔.
But I agree a `NotImplementedError` would be appropriate.

**Comment by twaclaw** [comment]
I can take a look at the linear algebra ops.

**Comment by ricardoV94** [comment]
Seems like Reshape is become more and more relevant, if anyone wants to tackle it

**Comment by twaclaw** [comment]
> Seems like Reshape is become more and more relevant, if anyone wants to tackle it

Is someone working on this?

**Comment by ricardoV94** [comment]
> > Seems like Reshape is become more and more relevant, if anyone wants to tackle it
> 
> Is someone working on this?

Not yet I think. You can go ahead

**Comment by Ch0ronomato** [comment]
Is the checklist at the top up to date on what else is needed?

**Comment by ricardoV94** [comment]
More or less up to date except linalg and indexing is being worked on

**Comment by ricardoV94** [comment]
If someone is interested we need to check whether we can bridge nicely between PyTensor and Torch random number generator APIs.

We have added a recent documentation page explaining how random variables work in PyTensor:

**Comment by Ch0ronomato** [comment]
@ricardoV94 I'll take a stab after i finish some of the operators in #939. I need to build a bit more familiarity with the pytensor code first

**Comment by danielgerardclaassen** [comment]
Coming from PyMC, adding a sparse solve would be useful I believe...

**Comment by ricardoV94** [comment]
> Coming from PyMC, adding a sparse solve would be useful I believe...

This issue is not very relevant for that request, since we first need it in PyTensor to begin with, before we add it to the PyTorch backend. We haven't done anything with Sparse stuff in the PyTorch backend to begin with

**Comment by aman-coder03** [comment]
hey @ricardoV94, i have been going through the codebase and would love to take a stab at `Blockwise`. I am thinking `torch.vmap` would be the right primitive here but wanted to check with you first before I go too far down that path, especially since you mentioned `torch.compile` compatibility as a goal. Does that approach sound reasonable, or are there gotchas I should know about upfront?
